### PR TITLE
Don't remove duplicate link libs

### PR DESCRIPTION
**Pull request by dg0yt**
Straight from the vcpkg port: Removing (late) duplicates breaks linking.

For static linkage, order of link libraries matters for most platforms. It is often okay to leave the last occurence, but it is not okay to leave the first occurence of a lib.

### CODE REVIEW

**Review by akallabeth — CHANGES_REQUESTED** [review]
we´re building fully static builds successfully on ci, so this is definitely not required.

**Comment by akallabeth** [comment]
@dg0yt I´ve just successfully build static binaries with `cmake -GNinja -Bbuild -S. -DCMAKE_INSTALL_PREFIX=/install -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release && cmake --build build --target install`
if you want it fully static check `ci/cmake-preloads/config-qa-static.cmake` for hints

**Comment by dg0yt** [comment]
I already do have a setup to build with static library linkage and dynamic runtime linkage.

And no matter how I would build:
If  liba needs libx, AND libb needs libx, then
~~~cmake
set(flags -la -lx)
list(APPEND flags -lb -lx)
list(REMOVE_DUPLICATES flags)
~~~
will leave libb with unsatisifed symbols, as in `-la -lx -lb`, while the valid reduction would be `-la -lb -lx`.
(This is extremely simplified. CMake orders link libs by (target) dependencies. Plain lists are not taken literally. Until they are eventually dumped into pkg-config `Libs` or similar.) 

This pattern has been seen a few times in packages, and the vcpkg freerdp patch was added when it was an issue. It is nice that is not an issue now, but it can pop up again when the configuration of dependencies changes.

File: libfreerdp/CMakeLists.txt [hunk 1]
```diff
@@ -430,7 +430,6 @@ endif()
 list(APPEND LIBFREERDP_PUB_LIBS winpr)
 
 list(REMOVE_DUPLICATES LIBFREERDP_DEFINITIONS)
-list(REMOVE_DUPLICATES LIBFREERDP_LIBS)
 list(REMOVE_DUPLICATES LIBFREERDP_PUB_LIBS)
```
Comment (dg0yt) [inline libfreerdp/CMakeLists.txt:433]: Oversight!

Comment (dg0yt) [inline libfreerdp/CMakeLists.txt:433]:
```suggestion
```

**Comment by dg0yt** [comment]
Details for the effect of `REMOVE_DUPLICATES`

Picking `LIBFREERDP_LIBS`, I see the following effect in my vcpkg builds (using shared X11 libs from the System):
~~~diff
-LIBFREERDP_LIBS=<vcpkg-root>/lib/libswscale.a;ZLIB::ZLIB;-lpthread;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;-lpthread;dl;/usr/lib/x86_64-linux-gnu/libSM.so;/usr/lib/x86_64-linux-gnu/libICE.so;/usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so;/usr/lib/x86_64-linux-gnu/libxkbfile.so;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;-lpthread;dl;<vcpkg-root>/lib/libavcodec.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libswresample.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libswresample.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;cjson;<vcpkg-root>/lib/libavcodec.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libswresample.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libswresample.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libavutil.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic
+LIBFREERDP_LIBS=<vcpkg-root>/lib/libswscale.a;ZLIB::ZLIB;-lpthread;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;dl;/usr/lib/x86_64-linux-gnu/libSM.so;/usr/lib/x86_64-linux-gnu/libICE.so;/usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so;/usr/lib/x86_64-linux-gnu/libxkbfile.so;<vcpkg-root>/lib/libavcodec.a;/usr/lib/x86_64-linux-gnu/libm.so;atomic;<vcpkg-root>/lib/libswresample.a;<vcpkg-root>/lib/libavutil.a;cjson
~~~
The spots of interest are e.g. `libswresample.a` and `libavutil.a` which came with trailing `m` and `atomic` but appear as independent in the result.

And now I have to admit that the input is the result of another workaround from past static linkage trouble: The freerdp FindFFmpeg uses pkg-config only to locate the root of ffmpeg, but ignores all other information which is already available in the pkg-config files. For pristine freerdp, I get:
~~~diff
-LIBFREERDP_LIBS=<vcpkg-root>/lib/libswscale.a;ZLIB::ZLIB;-lpthread;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;-lpthread;dl;/usr/lib/x86_64-linux-gnu/libSM.so;/usr/lib/x86_64-linux-gnu/libICE.so;/usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so;/usr/lib/x86_64-linux-gnu/libxkbfile.so;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;-lpthread;dl;<vcpkg-root>/lib/libavcodec.a;<vcpkg-root>/lib/libavutil.a;<vcpkg-root>/lib/libswresample.a;cjson;<vcpkg-root>/lib/libavcodec.a;<vcpkg-root>/lib/libavutil.a;<vcpkg-root>/lib/libswresample.a
+LIBFREERDP_LIBS=<vcpkg-root>/lib/libswscale.a;ZLIB::ZLIB;-lpthread;<vcpkg-root>/lib/libssl.a;<vcpkg-root>/lib/libcrypto.a;dl;/usr/lib/x86_64-linux-gnu/libSM.so;/usr/lib/x86_64-linux-gnu/libICE.so;/usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so;/usr/lib/x86_64-linux-gnu/libxkbfile.so;<vcpkg-root>/lib/libavcodec.a;<vcpkg-root>/lib/libavutil.a;<vcpkg-root>/lib/libswresample.a;cjson
~~~

I guess the build succeeds despite of `REMOVE_DUPLICATES` because the missing link libs (`atomic`, `m`, `pthread`) are added as a general pattern for some conditions. However, it won't work in general, and it requires you to burns transitive usage requirements into freerdp's CMake build system even when dependencies provide all information CMake needs to know.

**Comment by akallabeth** [comment]
@dg0yt as already mentioned, we build fully static stuff with the pull request builder and nightly (windows binaries are static, code quality checker does a linux static build)
so your patches look like configuration issues on your end to me (I may be wrong, but first would like to hear why our builds work and yours don´t )

**Comment by dg0yt** [comment]
> so your patches look like configuration issues on your end to me (I may be wrong, but first would like to hear why our builds work and yours don´t )

I already disclosed that even my current builds (3.3, 3.4) do work. And I still say my position is that this doesn't prove that `REMOVE_DUPLICATES` is right - it just proves that missing link libs are append in another way.

Please look at my previous explanation:

> If liba needs libx, AND libb needs libx, then
> 
> ```cmake
> set(flags -la -lx)
> list(APPEND flags -lb -lx)
> list(REMOVE_DUPLICATES flags)
> ```
> 
> will leave libb with unsatisifed symbols, as in `-la -lx -lb`, while the valid reduction would be `-la -lb -lx`. (This is extremely simplified. CMake orders link libs by (target) dependencies. Plain lists are not taken literally. Until they are eventually dumped into pkg-config `Libs` or similar.)

Isn't there an option to use cairo? Let me illustrate the usage requirements for cairo, based on my most recent work on `pkgconf`. If order isn't taken from pkgconf or cmake dependencies, you must duplicate that requirement information.
cairo

(Blue edges are dependencies, red edges are just topological sort. FTR my contributions to `pkgconf` were triggered by the fact that the pkgconf 2.x releases did print `--libs` in the right order for graphviz, the package generating this nice picture from pkgconf output. And graphviz uses pangocairo.)

**Review by akallabeth — APPROVED** [review]
Ok, you´ve convinced me.
should not be an issue except some deps showing up multiple times in pkg-config.
(hint: we do not use pkg-config for static builds, maybe that is the reason we did not have issues so far?)

**Comment by akallabeth** [comment]
@dg0yt you might want to squash the changes and check the other libraries for this too. (just grep for `REMOVE_DUPLICATES`)

**Comment by dg0yt** [comment]
Squashed. I don't want to touch any of the remaining:
~~~
$ git grep REMOVE_DUPLICATES
CMakeLists.txt: list(REMOVE_DUPLICATES CMAKE_FIND_ROOT_PATH)
channels/client/CMakeLists.txt: list(REMOVE_DUPLICATES CHANNEL_STATIC_CLIENT_ENTRIES)
cmake/MergeStaticLibs.cmake:                    list(REMOVE_DUPLICATES libfiles_${CONFIG_TYPE})
cmake/MergeStaticLibs.cmake:    list(REMOVE_DUPLICATES libfiles)
libfreerdp/CMakeLists.txt:list(REMOVE_DUPLICATES LIBFREERDP_DEFINITIONS)
libfreerdp/CMakeLists.txt:list(REMOVE_DUPLICATES LIBFREERDP_INCLUDES)
winpr/libwinpr/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_DEFINITIONS)
winpr/libwinpr/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_COMPILE_OPTIONS)
winpr/libwinpr/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_LINK_OPTIONS)
winpr/libwinpr/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_LINK_DIRS)
winpr/libwinpr/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_INCLUDES)
winpr/tools/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_TOOLS_DEFINITIONS)
winpr/tools/CMakeLists.txt:list(REMOVE_DUPLICATES WINPR_TOOLS_INCLUDES)
~~~

**Review by akallabeth — APPROVED** [review]
LGTM

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Pull request merged**